packages/core/src/features/st-structure.ts

```diff
@@ -0,0 +1,564 @@
+        'error',
+        (params: string) => `invalid @st "${params}" definition`
+    ),
+};
+
+export interface PartSymbol extends HasParts, STCustomState.HasStates {
+    _kind: 'part';
+    name: string;
+    id: string;
+    mapTo: ImmutableSelectorList | CSSClass.ClassSymbol;
+}
+export interface HasParts {
+    '-st-parts': Record<string, PartSymbol>;
+}
+
+export const experimentalMsg = '[experimental feature] stylable structure (@st): API might change!';
+
+const dataKey = plugableRecord.key<{
+    isStructureMode: boolean;
+    analyzedDefs: WeakMap<postcss.AtRule, AnalyzedStDef>;
+    analyzedDefToPartSymbol: Map<AnalyzedStDef, PartSymbol>;
+}>('st-structure');
+
+// HOOKS
+
+export const hooks = createFeature({
+    analyzeInit(context) {
+        const { meta } = context;
+        const stAtRule = meta.sourceAst.nodes.find((node) => isStAtRule(node));
+        if (stAtRule) {
+            warnOnce(experimentalMsg);
+            const metaAnalysis = plugableRecord.getUnsafe(context.meta.data, dataKey);
+            metaAnalysis.isStructureMode = true;
+        } else if (meta.type === 'stylable') {
+            // set implicit root for legacy mode (root with flat structure)
+            meta.root = 'root';
+            const rootSymbol = CSSClass.addClass(context, 'root');
+            rootSymbol[`-st-root`] = true;
+        }
+    },
+    metaInit({ meta }) {
+        plugableRecord.set(meta.data, dataKey, {
+            // default to legacy flat mode
+            isStructureMode: false,
+            analyzedDefs: new WeakMap(),
+            analyzedDefToPartSymbol: new Map(),
+        });
+    },
+    analyzeAtRule({ context, atRule, analyzeRule }) {
+        if (!isStAtRule(atRule)) {
+            return;
+        }
+
+        const { analyzedDefToPartSymbol } = plugableRecord.getUnsafe(context.meta.data, dataKey);
+        const analyzed = analyzeStAtRule(atRule, context);
+        if (!analyzed) {
+            if (atRule.parent?.type === 'root') {
+                context.diagnostics.report(diagnostics.UNSUPPORTED_TOP_DEF(), {
+                    node: atRule,
+                });
+            } else {
+                context.diagnostics.report(diagnostics.INVALID_ST_DEF(atRule.params), {
+                    node: atRule,
+                });
+            }
+        } else if (analyzed.type === 'topLevelClass') {
+            // ToDo: error when nested (only top level for now)
+            if (!analyzed.name) {
+                context.diagnostics.report(diagnostics.UNSUPPORTED_TOP_DEF(), {
+                    node: atRule,
+                });
+            }
+            const existingSymbol = STSymbol.get(context.meta, analyzed.name);
+            if (existingSymbol?._kind === 'import') {
+                context.diagnostics.report(diagnostics.OVERRIDE_IMPORTED_CLASS(), {
+                    node: atRule,
+                });
+                return;
+            }
+            // ToDo: pass atuRule for diagnostics
```

Comment:
todo

Reply:
Passed, but currently there are no actual diagnostics in the class symbol creation, so no tests.
Potentially there is the inner st-symbol re-declare diagnostic, but there refactoring the css-class registration is a bit of a sidetrack, so added it to the tech-debt to resolve after this PR.